src/network/tcpsocket.c

```diff
@@ -82,48 +82,37 @@ static void client_to_server_state(struct ustream *s) {
 static void client_read_cb(struct ustream *s, int bytes) {
     char *str;
     int len = 0;
-    uint32_t final_len;
-    int max_retry = 3, tried = 0;
-
-    do {
-        str = ustream_get_read_buf(s, &len);
-        if (!str)
-            break;
-
-        if (network_config.use_symm_enc) {
-            final_len = ntohl(*(uint32_t *)str);
-            if(len < final_len) {//not complete msg, wait for next recv
-                fprintf(stderr,"not complete msg, len:%d, expected len:%u\n", len, final_len);
-                if (tried++ == max_retry) {
-                    ustream_consume(s, len);
-                    return;//drop package
-                }
-                continue;
-            }
-            char *dec = gcrypt_decrypt_msg(str+sizeof(final_len), final_len-sizeof(final_len));
-        
-            handle_network_msg(dec);
-            free(dec);
-            ustream_consume(s, final_len);//one msg is processed
-            tried = 0;
-        } else {
-            final_len = ntohl(*(uint32_t *)str);
-            if(len < final_len){
-                fprintf(stderr,"not complete msg, len:%d, expected len:%u\n", len, final_len);
-                if (tried++ == max_retry) {
-                    ustream_consume(s, len);
-                    return;
-                }
-                continue;
-            } 
-            char* msg = malloc(final_len);
-            memcpy(msg, str+sizeof(final_len), final_len-sizeof(final_len));
-            handle_network_msg(msg);
-            ustream_consume(s, final_len);//one msg is processed
-            free(msg);
-            tried = 0;
+    uint32_t final_len = sizeof(uint32_t);
+    str = malloc(final_len);
+
+    if (ustream_read(s, str, final_len) < final_len) {//ensure recv sizeof(uint32_t).
```

Comment:
do u know what happens if a message is not completely? why does this happen? if u return, will the callback be called with the same message (but now complete) again, or is the message just discarded?

Reply:
usteam_read has a internal loop, keeping recv until error happens or received length requitrement is met.

345 int ustream_read(struct ustream *s, char *buf, int buflen)
346 {
347         char *chunk;
348         int chunk_len;
349         int len = 0;
350 
351         do {
352                 chunk = ustream_get_read_buf(s, &chunk_len);
353                 if (!chunk)
354                         break;
355                 if (chunk_len > buflen - len)
356                         chunk_len = buflen - len;
357                 memcpy(buf + len, chunk, chunk_len);
358                 ustream_consume(s, chunk_len);
359                 len += chunk_len;
360         } while (len < buflen);
361 
362         return len;
363 }

Consequently, the ustream_read function has the same usage as the original code.

Reply:
As a result, if one message is incomplete, it is discarded.

BTW, I don't think if the software is well coded, there would be an incomplete package in TCP send.

Since TCP has the ability to ensure the completeness.

This function is just an insurance to filter out absurd message sent by the current version of DAWN.

Reply:
This function could be used in another way that finding the bug of DAWN.
Since if a bug is encountered, it would break the older version of DAWN(no check before msg processed) and APs reboot log disappears. With this patch, such issue would not happen.

Reply:
the initial incompleteness would occur because the event driven loop would be triggered whenever one byte is recieved, not the complete message.